### ISSUE TITLE
Android:  Licence being ignored leadeing to Null Reference Exception

### ISSUE THREAD

**Issue by microd1** [adamfisher/Xamarin.Forms.VideoPlayer#97]
Hi,

I am using Octane.Xamarin.Forms.VideoPlayer in a Xamarin Forms solution targeting both iOS and Android.  I am using a local MP4 file.

My Android VideoPlayer implementation appears to be ignoring my license key.     When playing a video longer than 15 seconds, I get a warning:

`VideoPlayer [ERROR]: You cannot play a video longer than 15 seconds.01-27 17:07:09.797 I/mono-stdout(26092): You are attempting to use functionality that is not included in the trial version of this component. Please consider purchasing a licensed copy at https://components.xamarin.com/view/video-player.`

When the video runs to completion I get a NullReferenceException:

`System.NullReferenceException: Object reference not set to an instance of an object.`

`01-27 17:07:35.757 I/MonoDroid(26092):   at Octane.Xamarin.Forms.VideoPlayer.VideoPlayer+<>c.<.ctor>b__77_6 (System.Object sender, 
Octane.Xamarin.Forms.VideoPlayer.Events.VideoPlayerEventArgs args) [0x0001f] in <cbca696c09c24289b1d51447effb49e2>:0 `

`01-27 17:07:35.757 I/MonoDroid(26092):   at Octane.Xamarin.Forms.VideoPlayer.ExtensionMethods.EventHandlerExtensions.RaiseEvent[TEventArgs] (System.EventHandler1[TEventArgs] eventHandler, System.Object sender, TEventArgs eventArgs) [0x00003] in <cbca696c09c24289b1d51447effb49e2>:0 `

`01-27 17:07:35.757 I/MonoDroid(26092):   at Octane.Xamarin.Forms.VideoPlayer.VideoPlayer.OnCompleted (Octane.Xamarin.Forms.VideoPlayer.Events.VideoPlayerEventArgs e) [0x00000] in <cbca696c09c24289b1d51447effb49e2>:0 
...`

(I think the null reference exception is normal for videos longer than 15 seconds in trial mode)

- If I play a video that's shorter than 15 seconds,  the playback completes normally and there is no exception and everything's nice.
- If I stop the >15 seconds video before it gets to completion then the exception doesn't fire.
- My iOS implementation is working nicely without any issues.
- If I replace my license key with an invalid value I get a warning that my key is invalid.

I believe that the NullReferenceException is a result of the license key being ignored.

I'm using version 2.1.0

Please help!

**Comment by microd1** [comment]
My MainActivity.cs OnCreate method:

`protected override void OnCreate(Bundle bundle) {

           this.Window.AddFlags(WindowManagerFlags.Fullscreen);
            TabLayoutResource = Resource.Layout.Tabbar;
            ToolbarResource = Resource.Layout.Toolbar;

            base.OnCreate(bundle);
           
            global::Xamarin.Forms.Forms.Init(this, bundle);
            Octane.Xam.VideoPlayer.Android.FormsVideoPlayer.Init("F9redactedE4");
                                                                  
                Acr.UserDialogs.UserDialogs.Init(() => Plugin.CurrentActivity.CrossCurrentActivity.Current.Activity);

                //Hide the Navigation controls during play
                Window.DecorView.SystemUiVisibilityChange += visibilityListener;
                Window.DecorView.SystemUiVisibility = StatusBarVisibility.Hidden;

                AuditApp.Android.AndroidPlaystoreAudit.Instance.Run(this);

                LoadApplication(new MyGameShow.App());
        }
`

**Comment by microd1** [comment]
The debug log while the video is playing, showing the license error:

01-27 17:06:49.757 I/MediaPlayer(26092): sendBroadcast android.media.IMediaPlayer.videoexist
01-27 17:06:49.757 V/MediaPlayer-JNI(26092): start
01-27 17:06:49.757 V/MediaPlayer(26092): start
01-27 17:06:49.757 V/MediaPlayer-JNI(26092): getCurrentPosition: 0 (msec)
01-27 17:06:49.757 V/MediaPlayer(26092): getDuration_l
01-27 17:06:49.767 V/MediaPlayer-JNI(26092): getDuration: 25937 (msec)
01-27 17:06:49.777 I/MediaPlayer(26092): send context aware event
01-27 17:06:49.817 I/MediaPlayer(26092): sendBroadcast CONTEXT_AWARE_MUSIC_INFO - type(start) - id (60)
01-27 17:06:49.907 V/MediaPlayer(26092): message received msg=6, ext1=0, ext2=0
01-27 17:06:49.907 V/MediaPlayer(26092): Received MEDIA_STARTED
01-27 17:06:49.907 V/MediaPlayer(26092): callback application
01-27 17:06:49.907 V/MediaPlayer(26092): back from callback
01-27 17:06:49.907 V/MediaPlayer-JNI(26092): getCurrentPosition: 0 (msec)
01-27 17:06:49.917 V/MediaPlayer-JNI(26092): isPlaying: 1
01-27 17:06:49.967 V/MediaPlayer(26092): message received msg=5, ext1=640, ext2=480
01-27 17:06:49.967 V/MediaPlayer(26092): New video size 640 x 480
01-27 17:06:49.967 V/MediaPlayer(26092): callback application
01-27 17:06:49.967 V/MediaPlayer(26092): back from callback
01-27 17:06:49.967 V/MediaPlayer(26092): getVideoWidth
01-27 17:06:49.967 V/MediaPlayer-JNI(26092): getVideoWidth: 640
01-27 17:06:49.967 V/MediaPlayer(26092): getVideoHeight
01-27 17:06:49.967 V/MediaPlayer-JNI(26092): getVideoHeight: 480
01-27 17:06:49.967 V/MediaPlayer(26092): message received msg=200, ext1=3, ext2=0
01-27 17:06:49.967 W/MediaPlayer(26092): this is IMEDIA_PLAYER_VIDEO_EXIST
01-27 17:06:49.967 W/MediaPlayer(26092): info/warning (3, 0)
01-27 17:06:49.967 V/MediaPlayer(26092): callback application
01-27 17:06:49.967 V/MediaPlayer(26092): back from callback
01-27 17:06:49.987 I/MediaPlayer(26092): sendBroadcast android.media.IMediaPlayer.videoexist
01-27 17:06:50.757 V/MediaPlayer-JNI(26092): getCurrentPosition: 756 (msec)
01-27 17:06:51.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 1728 (msec)
01-27 17:06:52.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 2731 (msec)
01-27 17:06:53.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 3731 (msec)
01-27 17:06:54.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 4732 (msec)
01-27 17:06:54.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 4742 (msec)
01-27 17:06:54.747 V/MediaPlayer(26092): getDuration_l
01-27 17:06:54.747 V/MediaPlayer-JNI(26092): getDuration: 25937 (msec)
01-27 17:06:55.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 5732 (msec)
01-27 17:06:56.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 6731 (msec)
01-27 17:06:57.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 7731 (msec)
01-27 17:06:58.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 8732 (msec)
01-27 17:06:59.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 9732 (msec)
01-27 17:06:59.757 V/MediaPlayer-JNI(26092): getCurrentPosition: 9756 (msec)
01-27 17:06:59.757 V/MediaPlayer(26092): getDuration_l
01-27 17:06:59.757 V/MediaPlayer-JNI(26092): getDuration: 25937 (msec)
01-27 17:07:00.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 10733 (msec)
01-27 17:07:01.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 11734 (msec)
01-27 17:07:02.727 V/MediaPlayer-JNI(26092): getCurrentPosition: 12733 (msec)
01-27 17:07:03.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 13735 (msec)
01-27 17:07:04.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 14736 (msec)
01-27 17:07:04.757 V/MediaPlayer-JNI(26092): getCurrentPosition: 14756 (msec)
01-27 17:07:04.757 V/MediaPlayer(26092): getDuration_l
01-27 17:07:04.757 V/MediaPlayer-JNI(26092): getDuration: 25937 (msec)
01-27 17:07:05.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 15736 (msec)
01-27 17:07:06.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 16736 (msec)
01-27 17:07:07.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 17735 (msec)
01-27 17:07:08.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 18736 (msec)
01-27 17:07:09.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 19736 (msec)
01-27 17:07:09.757 V/MediaPlayer-JNI(26092): getCurrentPosition: 19756 (msec)
01-27 17:07:09.757 V/MediaPlayer(26092): getDuration_l
01-27 17:07:09.767 V/MediaPlayer-JNI(26092): getDuration: 25937 (msec)
01-27 17:07:09.797 I/mono-stdout(26092): VideoPlayer [ERROR]: You cannot play a video longer than 15 seconds.
VideoPlayer [ERROR]: You cannot play a video longer than 15 seconds.01-27 17:07:09.797 I/mono-stdout(26092): You are attempting to use functionality that is not included in the trial version of this component. Please consider purchasing a licensed copy at https://components.xamarin.com/view/video-player.

You are attempting to use functionality that is not included in the trial version of this component. Please consider purchasing a licensed copy at https://components.xamarin.com/view/video-player.
01-27 17:07:10.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 20737 (msec)
01-27 17:07:11.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 21737 (msec)
01-27 17:07:12.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 22736 (msec)
01-27 17:07:13.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 23739 (msec)
01-27 17:07:14.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 24737 (msec)
01-27 17:07:14.757 V/MediaPlayer-JNI(26092): getCurrentPosition: 24756 (msec)
01-27 17:07:14.757 V/MediaPlayer(26092): getDuration_l
01-27 17:07:14.757 V/MediaPlayer-JNI(26092): getDuration: 25937 (msec)
01-27 17:07:15.737 V/MediaPlayer-JNI(26092): getCurrentPosition: 25738 (msec)
01-27 17:07:15.897 V/MediaPlayer(26092): message received msg=2, ext1=0, ext2=0
01-27 17:07:15.897 V/MediaPlayer(26092): playback complete
01-27 17:07:15.897 V/MediaPlayer(26092): callback application
01-27 17:07:15.897 V/MediaPlayer(26092): back from callback
01-27 17:07:15.917 V/MediaPlayer-JNI(26092): getCurrentPosition: 25892 (msec)
01-27 17:07:15.917 V/MediaPlayer(26092): getDuration_l
01-27 17:07:15.917 V/MediaPlayer-JNI(26092): getDuration: 25937 (msec)
Unhandled Exception:

System.NullReferenceException: Object reference not set to an instance of an object.

**Comment by microd1** [comment]
The Null Reference Exception that happens if I get the license warning:

01-27 17:07:35.707 I/MonoDroid(26092): UNHANDLED EXCEPTION:
01-27 17:07:35.757 I/MonoDroid(26092): System.NullReferenceException: Object reference not set to an instance of an object.
01-27 17:07:35.757 I/MonoDroid(26092):   at Octane.Xamarin.Forms.VideoPlayer.VideoPlayer+<>c.<.ctor>b__77_6 (System.Object sender, Octane.Xamarin.Forms.VideoPlayer.Events.VideoPlayerEventArgs args) [0x0001f] in <cbca696c09c24289b1d51447effb49e2>:0 
01-27 17:07:35.757 I/MonoDroid(26092):   at Octane.Xamarin.Forms.VideoPlayer.ExtensionMethods.EventHandlerExtensions.RaiseEvent[TEventArgs] (System.EventHandler`1[TEventArgs] eventHandler, System.Object sender, TEventArgs eventArgs) [0x00003] in <cbca696c09c24289b1d51447effb49e2>:0 
01-27 17:07:35.757 I/MonoDroid(26092):   at Octane.Xamarin.Forms.VideoPlayer.VideoPlayer.OnCompleted (Octane.Xamarin.Forms.VideoPlayer.Events.VideoPlayerEventArgs e) [0x00000] in <cbca696c09c24289b1d51447effb49e2>:0 
01-27 17:07:35.757 I/MonoDroid(26092):   at Octane.Xamarin.Forms.VideoPlayer.Android.Renderers.VideoPlayerRenderer.<RegisterEvents>b__12_2 (System.Object sender, System.EventArgs args) [0x00017] in <b990b24acb5843e294728d2e4aac3e89>:0 
01-27 17:07:35.757 I/MonoDroid(26092):   at (wrapper delegate-invoke) <Module>:invoke_void_object_EventArgs (object,System.EventArgs)
01-27 17:07:35.757 I/MonoDroid(26092):   at Android.Media.MediaPlayer+IOnCompletionListenerImplementor.OnCompletion (Android.Media.MediaPlayer mp) [0x00011] in <28e323a707a2414f8b493f6d4bb27c8d>:0 
01-27 17:07:35.757 I/MonoDroid(26092):   at Android.Media.MediaPlayer+IOnCompletionListenerInvoker.n_OnCompletion_Landroid_media_MediaPlayer_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_mp) [0x0000f] in <28e323a707a2414f8b493f6d4bb27c8d>:0 
01-27 17:07:35.757 I/MonoDroid(26092):   at (wrapper dynamic-method) System.Object:7ec5765b-60b1-44e9-8f1a-edae2eff09d7 (intptr,intptr,intptr)
01-27 17:07:35.787 W/art     (26092): JNI RegisterNativeMethods: attempt to register 0 native methods for android.runtime.JavaProxyThrowable
01-27 17:07:35.787 D/Mono    (26092): DllImport searching in: '__Internal' ('(null)').
01-27 17:07:35.787 D/Mono    (26092): Searching for 'java_interop_jnienv_throw'.
01-27 17:07:35.787 D/Mono    (26092): Probing 'java_interop_jnienv_throw'.
01-27 17:07:35.787 D/Mono    (26092): Found as 'java_interop_jnienv_throw'.
An unhandled exception occured.

**Comment by adamfisher** [comment]
A couple of things you can try:

- Verify the bundle ID used in your `AndroidManifest.xml` correctly matches the one you requested a license for. This value is case sensitive and must match exactly.

- Usually cleaning the solution, uninstalling your app from the device/emulator, and cleaning bin and obj folders from your solution removes any caching. I have seen in the past where some users plug in a license key after using it in trial mode and for some reason Xamarin is performing some caching behind the scenes.

**Comment by microd1** [comment]
Hi Adam,

Thank you so much for responding to my issue!

Thanks for your clue about caching - it really helped me resolve a very weird issue:

While following your instructions to remedy Xamarain caching, I did a Google search for Xamarin Caching, just to see if there was any secret places other than Bin and Obj that Xamarin cached stuff.  I found an article that said that Xamarin did indeed cache stuff in the AppData folders, so I had a look in there.

I happened to find a Xamarin\Cache\Components\video-player-1.2.3.trial.xam and  Xamarin\Cache\Components\video-player-1.2.3.trial.info files in there.   "Strange", I thought, "These are left over trials from when Octane was distributed via Components instead of NuGet;  I wonder if they exist anywhere else?"

I searched for "video-player-" in my project folders and Lo and Behold!  Those two files were in a Components folder in my Project's folder. (Even though they were not listed in the Xamanrin Components list after being removed ages ago when I migrated to NuGet.)    I deleted them.

Visual Studio suddenly complained that I was referencing Octane.Xam.VideoPlayer which didn't exist - on the very line where I specify my license key!    I replaced that line with Octane.Xamarin.Forms.VideoPlayer and everything is now working properly!

So, what was happening was:  I was applying my license key to the Octane.**Xam** Video player shared instance (I don't know how it existed since I had uninstalled the component ages ago), but was playing the video with the Octane.**Xamarin.Forms** shared instance - which consequently had not been supplied with a license key.

Thanks so much for your hints that helped me resolve this very strange issue!
